pkg/workceptor/remote_work.go

```diff
@@ -0,0 +1,479 @@
+	return action(ctx, conn, reader)
+}
+
+// getConnectionAndRun retries connecting to a host and, once the connection succeeds, runs an action function.
+// If firstTimeSync is true, a single attempt is made on the calling goroutine. If the initial attempt fails to
+// connect or firstTimeSync is false, we run return ErrPending to the caller.
+func (rw *remoteUnit) getConnectionAndRun(ctx context.Context, firstTimeSync bool, action actionFunc) error {
+	if firstTimeSync {
+		err := rw.connectAndRun(ctx, action)
+		if err == nil {
+			return nil
+		} else if !utils.ErrorIsKind(err, "connection") {
+			return err
+		}
+	}
+	go func() {
+		conn, reader := rw.getConnection(ctx)
+		if conn != nil {
+			_ = action(ctx, conn, reader)
+		}
+	}()
+	return ErrPending
+}
+
+// startRemoteUnit makes a single attempt to start a remote unit.
+func (rw *remoteUnit) startRemoteUnit(ctx context.Context, conn net.Conn, reader *bufio.Reader) error {
+	closeOnce := sync.Once{}
+	doClose := func() error {
+		var err error
+		closeOnce.Do(func() {
+			err = conn.Close()
+		})
+		return err
+	}
+	defer doClose()
+	red := rw.Status().ExtraData.(*remoteExtraData)
+	_, err := conn.Write([]byte(fmt.Sprintf("work start %s\n", red.RemoteWorkType)))
+	if err != nil {
+		return fmt.Errorf("write error sending to %s: %s", red.RemoteNode, err)
+	}
+	response, err := reader.ReadString('\n')
+	if err != nil {
+		return fmt.Errorf("read error reading from %s: %s", red.RemoteNode, err)
+	}
+	submitIDRegex := regexp.MustCompile("with ID ([a-zA-Z0-9]+)\\.")
+	match := submitIDRegex.FindSubmatch([]byte(response))
+	if match == nil || len(match) != 2 {
+		return fmt.Errorf("could not parse response: %s", strings.TrimRight(response, "\n"))
+	}
+	red.RemoteUnitID = string(match[1])
+	rw.UpdateFullStatus(func(status *StatusFileData) {
+		ed := status.ExtraData.(*remoteExtraData)
+		ed.RemoteUnitID = red.RemoteUnitID
+	})
+	stdin, err := os.Open(path.Join(rw.UnitDir(), "stdin"))
+	if err != nil {
+		return fmt.Errorf("error opening stdin file: %s", err)
+	}
+	_, err = io.Copy(conn, stdin)
+	if err != nil {
+		return fmt.Errorf("error sending stdin file: %s", err)
+	}
+	err = doClose()
+	if err != nil {
+		return fmt.Errorf("error closing stdin file: %s", err)
+	}
+	response, err = reader.ReadString('\n')
+	if err != nil {
+		return fmt.Errorf("read error reading from %s: %s", red.RemoteNode, err)
+	}
+	resultErrorRegex := regexp.MustCompile("ERROR: (.*)")
+	match = resultErrorRegex.FindSubmatch([]byte(response))
+	if match != nil {
+		return fmt.Errorf("error from remote: %s", match[1])
+	}
+	rw.UpdateFullStatus(func(status *StatusFileData) {
+		ed := status.ExtraData.(*remoteExtraData)
+		ed.LocalStarted = true
+	})
+	return nil
+}
+
+// cancelOrReleaseRemoteUnit makes a single attempt to cancel or release a remote unit.
+func (rw *remoteUnit) cancelOrReleaseRemoteUnit(ctx context.Context, conn net.Conn, reader *bufio.Reader,
+	release bool, force bool) error {
+	closeOnce := sync.Once{}
```

Comment:
is this for future proofing? seems like we only have a single call to doClose() in this method

Reply:
No, it's this way because the function used to be more complicated and had multiple calls to doClose().  I'll simplify it.